hbase-server/src/main/java/org/apache/hadoop/hbase/replication/regionserver/CatalogReplicaReplicationEndpoint.java

```diff
@@ -0,0 +1,493 @@
+        .replay(regionInfo.getTable(), regionInfo.getEncodedNameAsBytes(), entries,
+          regionInfo.getReplicaId(), 1, operationTimeoutNs, location), (r, e) -> {
+        if (e != null) {
+          LOG.warn("Failed to replicate to {}", regionInfo);
+          LOG.warn("Exception: " + e);
+          future.completeExceptionally(e);
+        } else {
+          future.complete(r);
+        }
+      });
+
+      return future;
+    }
+
+    @Override public final void run() {
+      boolean reload = false;
+      boolean sleep = false;
+      List<Entry> entries = null;
+      while (isRunning()) { // we only loop back here if something fatal happened to our stream
+        if (location == null) {
+          // If location is null, keep searching for it. This applies to replica region server
+          // crashes, and waiting to be assigned. reload flag will be set accordingly in case of
+          // error.
+          try {
+            if (sleep) {
+              Thread.sleep(100);
+            }
+            location =
+              locator.getRegionLocation(regionInfo.getStartKey(), regionInfo.getReplicaId(), reload)
+                .get();
+            sleep = false;
+          } catch (ExecutionException ee) {
+            LOG.warn("Error getting location for meta region " + regionInfo + ", " + ee);
+            reload = true;
+            sleep = true;
+            continue;
+          } catch (InterruptedException ie) {
+            LOG.warn("Interrupted while getting location for meta region " +
+              regionInfo + ", " + ie);
+            Thread.currentThread().interrupt();
+          }
+        }
+        try {
+          if (entries == null) {
+            entries = queue.poll(20000, TimeUnit.MILLISECONDS);
+          }
+          if (entries == null) {
+            continue;
+          }
+
+          CompletableFuture<Long> future = replicate(entries);
+
+          try {
+            future.get();
+            entries = null;
+          } catch (InterruptedException e) {
+            // restore the interrupted state
+            Thread.currentThread().interrupt();
+            continue;
+          } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            LOG.warn("Failed to replicate {} entries for region  of meta table", entries.size());
+            LOG.warn("Cause: " + cause);
+            location = null;
+            reload = true;
+          }
+
+        } catch (InterruptedException | ReplicationRuntimeException e) {
+          // It is interrupted and needs to quit.
+          LOG.warn("Interrupted while waiting for next replication entry batch", e);
+          Thread.currentThread().interrupt();
+        }
+      }
+    }
+
+    /**
+     * @return whether the thread is running
+     */
+    public boolean isRunning() {
+      return isRunning && !isInterrupted();
+    }
+
+    /**
+     * @param running the running to set
+     */
+    public void setRunning(boolean running) {
```

Comment:
s/setRunning/ ... start/stop?

Reply:
Check if already running before setting flag?  Might have to synchronize the check/set or use atomicboolean

Reply:
Once the shipper is stopped, it cannot be restarted. Will change the method to stop(). I think there is no need to check the flag before setting if it is changed to stop(). volatile is good enough.

Reply:
Changed to stopShipper() as stop() cannot be overridden.